src/layers/attention/mask.rs

```diff
@@ -1,45 +1,50 @@
-use candle_core::Tensor;
-use snafu::{ResultExt, Snafu};
+use candle_core::{DType, IndexOp, Tensor};
+use snafu::{ensure, ResultExt, Snafu};
 
-/// Errors for attention masks.
 #[derive(Debug, Snafu)]
 pub enum AttentionMaskError {
     #[snafu(display("Cannot apply logits mask"))]
     ApplyLogitsMask { source: candle_core::Error },
 
     #[snafu(display("Cannot intersect masks"))]
     IntersectMasks { source: candle_core::Error },
-
-    #[snafu(display("Cannot reshape 2D mask to 4D"))]
-    ReshapeMask { source: candle_core::Error },
 }
 
 /// Attention mask.
 ///
-/// Sequence elements for which the corresponding mask element is set to
-/// `False` are ignored during attention calculation.
+/// A 4D attention mask with shape *(batch_size, heads, query_len, key_len)*.
+/// Elements for which the corresponding mask element is set to `False` are
+/// ignored during attention calculation.
 #[derive(Clone, Debug)]
 pub struct AttentionMask {
     bool_mask: Tensor,
 }
 
-impl AttentionMask {
-    /// Create an attention mask.
-    ///
-    /// * `bool_mask` - Boolean mask tensor.
-    ///   *Shape:* `(batch_size, seq_len)`
-    pub fn new(bool_mask: Tensor) -> Result<Self, AttentionMaskError> {
-        let bool_mask = match bool_mask.shape().dims2() {
-            Ok((batch_len, key_len)) => bool_mask
-                .reshape((batch_len, 1, 1, key_len))
-                .context(ReshapeMaskSnafu)?,
-            _ => bool_mask,
-        };
+impl From<InputMask> for AttentionMask {
+    fn from(input_mask: InputMask) -> Self {
+        AttentionMask::from(&input_mask)
+    }
+}
 
-        Ok(AttentionMask { bool_mask })
+impl From<&InputMask> for AttentionMask {
```

Comment:
Wouldn't it be better to implement `TryFrom` instead of panicking?

Reply:
It was `TryFrom` first, but the errors cannot happen because the invariants are guaranteed by `InputMask`. If these errors are tripped, we should panic, because it's an internal state/programming error and not a user error.